gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,146 @@
+    """``Geoclue`` was unable to retrieve the location due to a generic error."""
+
+    DENIED = auto()
+    """Permission to read the location was denied."""
+
+
+class Location(GObject.Object):
+    # Implement as GObject.Object to allow notifying on state changes within the event loop
+    state = GObject.Property(type=int, default=States.STARTING)
+
+    def __init__(self, interface):
+        super().__init__()
+
+        self.interface = interface
+
+        self.geoclue_simple = None
+        self.notify_location_handle = None
+        self.notify_active_listener = None
+        self.in_flatpak = None
+
+        def new_finish(_, async_result):
+            try:
+                self.geoclue_simple = Geoclue.Simple.new_finish(async_result)
+            except GLib.Error as e:
+                if e.matches(Gio.DBusError, Gio.DBusError.ACCESS_DENIED):
+                    # TODO: In practice, I cannot get this to occur. Maybe it is KDE's portal
+                    # implementation, but when I deny location permissions in the prompt,
+                    # the failure is a generic DBusError.FAILED, not a permission denied
+                    # failure.
+                    self.props.state = States.DENIED
+                else:
+                    self.props.state = States.FAILED
+                    print("Failed to get location", e.message, e.code)
+
+                return
+            else:
+                self.props.state = States.READY
+
+            client = self.geoclue_simple.get_client()
+            # Geoclue docs indicate a client proxy is not used in Flatpak
+            # https://lazka.github.io/pgi-docs/#Geoclue-2.0/classes/Simple.html#Geoclue.Simple.props.client
+            self.in_flatpak = client is None
+
+            if client:
+
+                def notify_client_active(*args):
+                    if not self.geoclue_simple.get_client().props.active:
+                        # If the client explicitly becomes inactive, this indicates
+                        # a failure to retrieve the location
+                        self.props.state = States.DENIED
+
+                self.notify_active_listener = client.connect(
+                    "notify::active", notify_client_active
+                )
+
+        Geoclue.Simple.new(
+            self.interface.app.app_id, Geoclue.AccuracyLevel.EXACT, None, new_finish
+        )
+
+    @property
+    def can_get_location(self):
+        return self.props.state in (States.READY, States.MONITORING)
+
+    def has_permission(self):
+        # TODO: Think this through a bit more
+        if self.props.state == States.STARTING:
+            # Cannot know whether permission request is needed...
+            # False is safe?
+            return False
+
+        if self.in_flatpak and not self.can_get_location:
+            # In flatpak, explicit permission required
+            return False
+
+        # Otherwise, no explicit permission required
+        return True
```

Comment:
This can probably be reduced to `return self.can_get_location`, I'm just not sure whether a distinction needs to clearly be made of when permission's been requested already or if it can/should be requested... or how that distinction is actually communicated. The doc string on the core Location object doesn't make this totally clear. Comes down to a bit of uncertainty on my part on the Location interface, and I probably just need to do a bit more reading of the other implementations.

Reply:
`has_location` is a strict boolean - if the platform requires permission, and you haven't asked, you don't have permission, `has_permission` returns False. `request_permission` will always trigger a hard answer to the question.

Internally, some of the platforms maintain a "unknown" status because it allows you to optimise the number of requests sent to the user - but that's an internal detail.